expression/expression.go

```diff
@@ -542,6 +542,9 @@ func EvaluateExprWithNull(ctx sessionctx.Context, schema *Schema, expr Expressio
 		for i, arg := range x.GetArgs() {
 			args[i] = EvaluateExprWithNull(ctx, schema, arg)
 		}
+		if x.FuncName.L == ast.Cast {
+			return NewFunctionInternal(ctx, x.FuncName.L, x.RetType, args...)
+		}
 		return NewFunctionInternal(ctx, x.FuncName.L, types.NewFieldType(mysql.TypeTiny), args...)
```

Comment:
Did you try to only change this line to `NewFunctionInternal(ctx, x.FuncName.L, x.RetType, args...)`?

Reply:
I've tried, and it can pass all the tests :)
However, according to [this line](https://github.com/pingcap/tidb/blob/master/expression/scalar_function.go#L141), it is possible that RetType cannot be overwritten by builtinRetTp. I don't know the exact reason why using mysql.TypeTiny as the default type, but in order to reduce the potential risks, I kept the original logic, that is, use mysql.TypeTiny as the default type.

Reply:
That line won't affect `x.RetType`.
`*retType = ...` would affect.

Reply:
In [this scalar function](https://github.com/pingcap/tidb/blob/master/expression/scalar_function.go#L144), when f.getRetTp() is mysql.TypeUnspecified, sf.RetType is mysql.TypeTiny, instead of mysql.TypeUnspecified.

If we change [this line](https://github.com/pingcap/tidb/blob/master/expression/expression.go#L545) to `NewFunctionInternal(ctx, x.FuncName.L, x.RetType, args...)`, sf.RetType may not be mysql.TypeTiny anymore when f.getRetTp() is mysql.TypeUnspecified. I'm not sure if this change has bad influence.

Reply:
This method just subsitute some columns with NULLs. Use original type should not be worse.

Reply:
Applied, thanks!